src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php

```diff
@@ -64,33 +99,116 @@ public function update(Response $response)
+
+        if (isset($flags['must-revalidate'])) {
+            $flags['no-cache'] = true;
         }
 
-        if (!$response->isFresh() || !$response->isCacheable()) {
-            $this->cacheable = false;
+        $response->headers->set('Cache-Control', implode(', ', array_keys($flags)));
+
+        $maxAge = null;
+        $sMaxage = null;
+
+        if ($this->ageDirectives['max-age']) {
+            $maxAge = $this->ageDirectives['max-age'] + $this->age;
+            $response->headers->addCacheControlDirective('max-age', $maxAge);
         }
 
-        if (!$this->cacheable) {
-            $response->headers->set('Cache-Control', 'no-cache, must-revalidate');
+        if ($this->ageDirectives['s-maxage']) {
+            $sMaxage = $this->ageDirectives['s-maxage'] + $this->age;
 
-            return;
+            if ($maxAge !== $sMaxage) {
+                $response->headers->addCacheControlDirective('s-maxage', $sMaxage);
+            }
+        }
+
+        if ($this->ageDirectives['expires']) {
+            $response->setExpires($response->getDate()->modify('+'.$this->ageDirectives['expires'].' seconds'));
         }
+    }
 
-        $this->ttls[] = $response->getTtl();
-        $this->maxAges[] = $response->getMaxAge();
+    /**
+     * RFC2616, Section 13.4.
+     *
+     * @see https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.4
+     *
+     * @return bool
+     */
+    private function willMakeFinalResponseUncacheable(Response $response)
+    {
+        // RFC2616: A response received with a status code of 200, 203, 300, 301 or 410
+        // MAY be stored by a cache […] unless a cache-control directive prohibits caching.
+        if ($response->headers->hasCacheControlDirective('no-cache')
+            || $response->headers->getCacheControlDirective('no-store')
+        ) {
+            return true;
+        }
 
-        if ($this->isNotCacheableResponseEmbedded) {
-            $response->headers->removeCacheControlDirective('s-maxage');
-        } elseif (null !== $maxAge = min($this->maxAges)) {
-            $response->setSharedMaxAge($maxAge);
-            $response->headers->set('Age', $maxAge - min($this->ttls));
+        // Last-Modified and Etag headers cannot be merged, they render the response uncacheable
+        // by default (except if the response also has max-age etc.).
+        if (\in_array($response->getStatusCode(), array(200, 203, 300, 301, 410))
+            && null === $response->getLastModified()
+            && null === $response->getEtag()
+        ) {
+            return false;
+        }
+
+        // RFC2616: A response received with any other status code (e.g. status codes 302 and 307)
+        // MUST NOT be returned in a reply to a subsequent request unless there are
+        // cache-control directives or another header(s) that explicitly allow it.
+        $cacheControl = array('max-age', 's-maxage', 'must-revalidate', 'proxy-revalidate', 'public', 'private');
```

Comment:
does must-revalidate indeed work as intended? that is, does the HttpCache on such a request also check all embedded ESI requests?

Reply:
It does not, and it cannot, at least sure not for ETag, but Last-Modified would be **very** complex as well. This list is taken from the RFC, not from "what might be useful in our case". Symfony always [adds "must-revalidate"](https://github.com/symfony/symfony/blob/77ae8bc10240721c44a78e3338d5c28dd7c11c55/src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php#L116-L118) [by default](https://github.com/symfony/symfony/blob/77ae8bc10240721c44a78e3338d5c28dd7c11c55/src/Symfony/Component/HttpFoundation/ResponseHeaderBag.php#L318), even though that can never be used because of missing validation headers. But @stof and I agreed to not change the existing behavior.

Reply:
i read up what must-revalidate is, and its about the handling of stale cache, so when max-age is exceeded. i somehow thought it forces to always use etag/last-modified, but that would be achieved by setting max-age 0 and an etag. sorry for the confusion here.